modules/core/core-elasticsearch/src/main/java/com/enonic/xp/elasticsearch/impl/ElasticsearchClientActivator.java

```diff
@@ -0,0 +1,182 @@
+package com.enonic.xp.elasticsearch.impl;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Hashtable;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.elasticsearch.action.admin.cluster.health.ClusterHealthRequest;
+import org.elasticsearch.action.admin.cluster.health.ClusterHealthResponse;
+import org.elasticsearch.cluster.health.ClusterHealthStatus;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.enonic.xp.elasticsearch.client.impl.EsClient;
+
+@Component(immediate = true, configurationPid = "com.enonic.xp.elasticsearch")
+public final class ElasticsearchClientActivator
+{
+
+    private static final Logger LOG = LoggerFactory.getLogger( ElasticsearchClientActivator.class );
+
+    private EsClient client;
+
+    private ServiceRegistration<EsClient> clientReg;
+
+    private BundleContext context;
+
+    private ScheduledExecutorService activateExecutorService;
+
+    private volatile boolean isRegistered = false;
```

Comment:
in code it looks like isRegistered = true only when clientReg is not null. you better synchronize register/unregister methods and change checking isRegistered to clientReg != null